File: C#/Dailies/Palindrome Checker/Advance.cs
/*
[Advanced]
     Rules:
Handles more complex cases and special characters.
Ignores spaces, punctuation, and letter casing while determining if the string is a palindrome.
*/

using System;
using System.Text.RegularExpressions;

class Program
{
    static void Main()
    {
        // Define the input string to check for palindrome
        string input = "A man, a plan, a canal, Panama!";

        // Check if the input string is a palindrome
        bool isPalindrome = IsPalindrome(input);

        // Display the result based on whether the input string is a palindrome or not
        if (isPalindrome)
        {
            Console.WriteLine("The input string is a palindrome.");
        }
        else
        {
            Console.WriteLine("The input string is not a palindrome.");
        }
    }

    // Function to check if a string is a palindrome
    static bool IsPalindrome(string input)
    {
        // Remove spaces, punctuation, and convert all characters to lowercase
        string cleanedString = CleanString(input);

        // Initialize pointers for checking palindrome
        int left = 0;
        int right = cleanedString.Length - 1;

        // Loop until the pointers meet or cross
        while (left < right)
        {
            // Compare characters at the left and right pointers
            if (cleanedString[left] != cleanedString[right])
            {
                // If characters don't match, the string is not a palindrome
                return false;
            }

            // Move the pointers towards the center
            left++;
            right--;
        }

        // If the loop completes without finding any mismatch, the string is a palindrome
        return true;
    }

    // Function to clean the input string by removing spaces, punctuation, and converting to lowercase
    static string CleanString(string input)
    {
        // Remove anything that's not a letter or a number and convert to lowercase
        return Regex.Replace(input, @"[^a-zA-Z0-9]", "").ToLower();
    }
}

File: Python/Dailies/Palindrome Checker/Advance.py
import re

# Function to check if a string is a palindrome
def is_palindrome_string(input_string):
    # Clean the input string by removing spaces, punctuation, and converting to lowercase
    cleaned_string = clean_string(input_string)

    # Check if the cleaned string is equal to its reverse
    return cleaned_string == cleaned_string[::-1]

# Function to clean the input string by removing spaces, punctuation, and converting to lowercase
def clean_string(input_string):
    # Remove anything that's not a letter or a number and convert to lowercase
    return re.sub(r'[^a-zA-Z0-9]', '', input_string.lower())


input_string = "A man, a plan, a canal, Panama!"

# Check if the input string is a palindrome
is_palindrome = is_palindrome_string(input_string)

# Display the result based on whether the input string is a palindrome or not
if is_palindrome:
    print("The input string is a palindrome.")
else:
    print("The input string is not a palindrome.")
    
